src/exchanges/kucoin-client.js

```diff
@@ -0,0 +1,312 @@
+const BasicClient = require("../basic-client");
+const Ticker = require("../ticker");
+const Trade = require("../trade");
+const Level2Point = require("../level2-point");
+const Level2Update = require("../level2-update");
+const Level2Snapshot = require("../level2-snapshot");
+const https = require("../https");
+const UUID = require("uuid/v4");
+const semaphore = require("semaphore");
+const { wait } = require("../util");
+
+class KucoinClient extends BasicClient {
+  constructor() {
+    super(undefined, "KuCoin");
+
+    this.hasTickers = true;
+    this.hasTrades = true;
+    this.hasLevel2Snapshots = false;
+    this.hasLevel2Updates = true;
+  }
+
+  _beforeConnect() {
+    this._wss.prependListener("connected", this._resetSemaphore.bind(this));
+    this._wss.on("connected", this._startPing.bind(this));
+    this._wss.on("disconnected", this._stopPing.bind(this));
+    this._wss.on("closed", this._stopPing.bind(this));
+  }
+
+  _resetSemaphore() {
+    this._sem = semaphore(1);
+  }
+
+  _startPing() {
+    clearInterval(this._pingInterval);
+    this._pingInterval = setInterval(this._sendPing.bind(this), 15000);
```

Comment:
We can probably use the `pingInterval` value returned in the HTTP request instead of hard coding this to 15s.